README.md
[![GoDoc](https://godoc.org/github.com/sjsafranek/go-mapnik?status.png)](https://godoc.org/github.com/sjsafranek/go-mapnik)
[![Go Report Card](https://goreportcard.com/badge/github.com/sjsafranek/go-mapnik)](https://goreportcard.com/report/github.com/sjsafranek/go-mapnik)
[![Version 0.1.4](https://img.shields.io/badge/version-1.4-brightgreen.svg)](http://sjsafranek.github.io/go-mapnik)

go-mapnik
=========

Go bindings for mapnik 2.2 and Mapnik 3.0 (http://www.mapnik.org or
http://github.com/mapnik/mapnik)

These bindings rely on http://github.com/springmeyer/mapnik-c-api.

Requires
-----------
### Postgresql 9.5



### PostgreSQL Setup (Linux):
#### Create new user
  `$ adduser mapnik`
  *** password 'dev'

#### create new postgres user and database table
$ sudo -i -u postgres
$ psql
 - postgres=# CREATE USER mapnik WITH PASSWORD 'dev';
 - postgres=# CREATE DATABASE mbtiles;
 - postgres=# GRANT ALL PRIVILEGES ON DATABASE mbtiles TO mapnik;

#### Check setup
$ sudo -i -u mapnik
$ psql -d mbtiles -U mapnik


### Run with PostgreSQL
config.json:
`{
  "cache": "postgres://mapnik:dev@localhost/mbtiles",
  "engine": "postgres",
  "layers": {
    "default": "sampledata/world/stylesheet.xml",
    "sample": "sampledata/world/stylesheet.xml",
    "population": "sampledata/world_population/population.xml",
    "osm": "https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
  },
  "port": 8080
}`

  `$ ./bin/tileserver -c config.json`


### Run with Sqlite3
config.json:
`{
  "cache": "tilecache.mbtiles",
  "engine": "sqlite",
  "layers": {
    "default": "sampledata/world/stylesheet.xml",
    "sample": "sampledata/world/stylesheet.xml",
    "population": "sampledata/world_population/population.xml",
    "osm": "https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
  },
  "port": 8080
}`

  `$ ./bin/tileserver -c config.json`


su - mapnik
sudo -i -u mapnik





Installation
-----------

### Linux / MacOS

1. Install Mapnik build environmnent,
	- e.g. Ubuntu: `apt-get install libmapnik-dev`
2. Download the sources, either by
    - `git clone` the repository to $GOPATH/src/fawick/go-mapnik

	OR

    - `go get -d github.com/fawick/go-mapnik/mapnik`
3. `cd mapnik` and run the configuration script `./configure.bash`.
   That script will setup the correct paths for including Mapnik headers and
   linking against the Mapnik shared library, as well as download the Mapnik C
   API source and `go install` the bindings.



### Windows

On Windows, go-mapnik is restricted to GOARCH=386 and Mapnik 2.2 for the moment,
as the precompiled 2.2 SDK is build with MSVC 32-bit and the C api must be build with a
compatible compiler as a DLL before the Go bindings can be linked against it.

You also need to have the MSVC 10 compiler installed on your system.

So, make sure `go version` reports `windows/386` as your toolchain. If you
happen to have the `windows/amd64` toolchain, either download Go binaries for
`windows/386` or put a MinGW 32-bit compiler in your path and rebuild the Go
binaries on your own.


1. Install Mapnik 2.2 Windows 32bit SDK from http://mapnik.org/pages/downloads.html
2. Download the sources, either by
    + `git clone` the repository to $GOPATH/src/fawick/go-mapnik

    OR

    + `go get -d github.com/fawick/go-mapnik/mapnik`
3. Run `configure.cmd` in the folder `mapnik` to compile a C DLL
   that can be used by Go/CGO/GCC later (sources will be downloaded
   automatically). Also, the script will  `go install` the bindings.
4. Run `go run demo.go` and open `view_tileserver.html` in a browser.
   (Make sure your %PATH% environment variable contains the paths of both
    `mapnik.dll` and the newly created `mapnik_c_api.dll`.)

Usage
-----

See `demo.go` for some usage examples.



MBTiles
-------

https://github.com/mapbox/mbtiles-spec/blob/master/1.2/spec.md



Related Work
------------

There is another Go package that offers access to mapnik at
https://github.com/omniscale/go-mapnik by Oliver Tonnhofer of Omniscale.
According to them, it is inspired by/based on this package. Instead of fetching
the latest mapnik-c-api they vendor their own c-code so, their package supports
some more features of libmapnik directly, such as version information access
and logging. It has support for Mapnik 3.0 built within.


CHANGELOG.md
Change Log
=========

## [Unreleased]


## [0.1.6] - 2017-04-07
### Added
 - restapi route for adding and getting tilelayers
### Changed
 - Read tilelayers from database


## [0.1.5] - 2017-04-03
### Added
 - Global variables
### Changed
 - merged duplicated functions for easier testing and debugging
 - logging reports server name and version

## [0.1.4] - 2017-04-03
### Added
 - Commenting for documentation
 - metadata route for each layer
### Changed
 - gorilla mux used for postgresql and sqlite3 servers
 - Postgres descriptions to table and fields
 - Updated Makefile
 - tms xml responses
### Fixed
 - metadata table support for multiple tile layers

## [0.1.3] - 2017-12-01
### Added
 - config file
 - tile url proxy
 - Vagrantfile

## [0.1.2] - 2016-11-01
### Added
 - RESTApi
 - Logging library
### Changed
 - Postgres backend option

## [0.1.1] - 2016-11-01
### Added
 - Fork from https://github.com/fawick/go-mapnik


tileserver/main.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"time"
)

import "maptiles"

type Config struct {
	Cache  string            `json:"cache"`
	Engine string            `json:"engine"`
	Layers map[string]string `json:"layers"`
	Port   int               `json:"port"`
}

var (
	config Config
	// engine string
	// port string
	// db_cache string
	config_file   string
	print_version bool
)

// Serve a single stylesheet via HTTP. Open view_tileserver.html in your browser
// to see the results.
// The created tiles are cached in an sqlite database (MBTiles 1.2 conform) so
// successive access a tile is much faster.
func TileserverWithCaching(engine string, layer_config map[string]string) {
	bind := fmt.Sprintf("0.0.0.0:%v", config.Port)
	if engine == "postgres" {
		t := maptiles.NewTileServerPostgresMux(config.Cache)

		// for i := range layer_config {
		// 	t.AddMapnikLayer(i, layer_config[i])
		// }

		maptiles.Ligneous.Info("Connecting to postgres database:")
		maptiles.Ligneous.Info("*** ", config.Cache)
		maptiles.Ligneous.Info(fmt.Sprintf("Magic happens on port %v...", config.Port))
		srv := &http.Server{
			Addr:         bind,
			Handler:      t.Router,
			ReadTimeout:  5 * time.Second,
			WriteTimeout: 10 * time.Second,
		}
		maptiles.Ligneous.Error(srv.ListenAndServe())
	} else {
		t := maptiles.NewTileServerSqliteMux(config.Cache)

		// for i := range layer_config {
		// 	t.AddMapnikLayer(i, layer_config[i])
		// }

		maptiles.Ligneous.Info("Connecting to sqlite3 database:")
		maptiles.Ligneous.Info("*** ", config.Cache)
		maptiles.Ligneous.Info(fmt.Sprintf("Magic happens on port %v...", config.Port))
		srv := &http.Server{
			Addr:         bind,
			Handler:      t.Router,
			ReadTimeout:  5 * time.Second,
			WriteTimeout: 10 * time.Second,
		}
		maptiles.Ligneous.Error(srv.ListenAndServe())
	}
}

func init() {
	// TODO: add config file
	// flag.StringVar(&port, "p", "8080", "server port")
	// flag.StringVar(&engine, "e", "sqlite", "database engine [sqlite or postgres]")
	// flag.StringVar(&db_cache, "d", "tilecache.mbtiles", "tile cache database")
	flag.StringVar(&config_file, "c", "", "tile server config")
	flag.BoolVar(&print_version, "v", false, "version")
	flag.Parse()
	// if engine != "sqlite" {
	// 	if engine != "postgres" {
	// 		logger.Fatal("Unsupported database engines")
	// 	}
	// }
	if print_version {
		fmt.Println(maptiles.SERVER_NAME + "-" + maptiles.VERSION)
		os.Exit(1)
	}
}

func getConfig() {
	// check if file exists!!!
	if _, err := os.Stat(config_file); err == nil {

		file, err := ioutil.ReadFile(config_file)
		if err != nil {
			panic(err)
		}

		err = json.Unmarshal(file, &config)
		if err != nil {
			fmt.Println("error:", err)
			os.Exit(1)
		}

		if config.Engine != "sqlite" {
			if config.Engine != "postgres" {
				fmt.Println("Unsupported database engine")
				os.Exit(1)
			}
		}

		maptiles.Ligneous.Debug(config)
	} else {
		fmt.Println("Config file not found")
		os.Exit(1)
	}
}

// Before uncommenting the GenerateOSMTiles call make sure you have
// the necessary OSM sources. Consult OSM wiki for details.
func main() {
	getConfig()
	TileserverWithCaching(config.Engine, config.Layers)
}

// sudo su mapnik
// psql -d mbtiles -U mapnik -W
